README.md
statsd-client
=============

[![Build Status](https://travis-ci.org/swoop-inc/statsd-client.svg?branch=swoop)](https://travis-ci.org/swoop-inc/statsd-client/builds) (**`swoop` branch status**)

A statsd client library implemented in Java. Allows for Java applications to easily communicate with statsd.

This version was originally forked from [DataDog/java-dogstatsd-client](https://github.com/DataDog/java-dogstatsd-client), [java-dogstatsd-client](https://github.com/indeedeng/java-dogstatsd-client) and [java-statsd-client](https://github.com/youdevise/java-statsd-client) but it is now the canonical home for swoop's statsd client.

See [CHANGELOG.md](CHANGELOG.md) for changes.

Maven Integration
-----------------

```xml
<dependency>
    <groupId>com.swoop</groupId>
    <artifactId>statsd-client</artifactId>
    <version>${swoop.statsd-client.version}</version>
</dependency>
```

Releases can be found here: <https://github.com/Shopximity/swoop-rest/releases>

Usage
-----
```java
import com.timgroup.statsd.ServiceCheck;
import com.timgroup.statsd.StatsDClient;
import com.timgroup.statsd.NonBlockingStatsDClient;

public class Foo {

  private static final StatsDClient statsd = new DisruptorStatsDClient(
    "my.prefix",                          /* prefix to any stats; may be null or empty string */
    "statsd-host",                        /* common case: localhost */
    8125,                                 /* port */
    new String[] {"tag:value"}            /* Datadog extension: Constant tags, always applied */
  );

  public static final void main(String[] args) {
    statsd.incrementCounter("foo");
    statsd.recordGaugeValue("bar", 100);
    statsd.recordGaugeValue("baz", 0.01); /* DataDog extension: support for floating-point gauges */
    statsd.recordHistogramValue("qux", 15);     /* DataDog extension: histograms */
    statsd.recordHistogramValue("qux", 15.5);   /* ...also floating-point */

    ServiceCheck sc = ServiceCheck
          .builder()
          .withName("my.check.name")
          .withStatus(ServiceCheck.Status.OK)
          .build();
    statsd.serviceCheck(sc); /* Datadog extension: send service check status */

    /* Compatibility note: Unlike upstream statsd, DataDog expects execution times to be a
     * floating-point value in seconds, not a millisecond value. This library
     * does the conversion from ms to fractional seconds.
     */
    statsd.recordExecutionTime("bag", 25, "cluster:foo"); /* DataDog extension: cluster tag */
  }
}
```


CHANGELOG.md
# CHANGELOG

Swoop Fork
==========
* Refactored for Java 8 only.
* Refactored codebase and APIs to be more extendable and DRY.
  * Leveraging Java 8 default interface methods to simplify implementations.
  * `StringMessageStatsDClient` abstract class now makes it easier to create custom clients. Corresponding test class can also be extended.
  * `NonBlockingStatsDClient` now extends from `StringMessageStatsDClient`.
  * `StatsDClient` now extends [`Closeable`](https://docs.oracle.com/javase/7/docs/api/java/io/Closeable.html) for easier use with Java 8's try-with-resources.
* Implemented LMAX disruptor version extending from `StringMessageStatsDClient`.

2.3 / 2016.10.21
================
* [BUGFIX] Reduce packet size to avoid fragmentation, see [#17](https://github.com/DataDog/java-dogstatsd-client/pull/17).
* [BUGFIX] Prefix sample rate with an '@', see [#15](https://github.com/DataDog/java-dogstatsd-client/pull/15).
* [FEATURE] Add support for sending `sample_rate` while submitting metrics, see [#11](https://github.com/DataDog/java-dogstatsd-client/pull/11).

2.2 / 2016.07.13
================
* [OTHER] Merged upstream from `indeedeng/java-dogstatsd-client`

2.1.1 / 2015.04.27
==================
* [BUGFIX] Constant tag support for service checks. See [#3][], thanks [@PatrickAuld][]
* [OTHER] Changelog update

2.1.0 / 2015.03.12
==================
Fork from [indeedeng/java-dogstatsd-client] (https://github.com/indeedeng/java-dogstatsd-client/releases/tag/java-dogstatsd-client-2.0.7) statsd client library
* [FEATURE] Add support for Datadog service checks


<!--- The following link definition list is generated by PimpMyChangelog --->
[#3]: https://github.com/DataDog/java-dogstatsd-client/issues/3
[@PatrickAuld]: https://github.com/PatrickAuld


src/test/java/com/timgroup/statsd/DummyStatsDServer.java

package com.timgroup.statsd;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.SocketException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;


class DummyStatsDServer {
    private final BlockingQueue<String> messagesReceived = new LinkedBlockingQueue<>();
    private final DatagramSocket server;

    public DummyStatsDServer(int port) throws SocketException {
        server = new DatagramSocket(port);
        Thread thread = new Thread(() -> {
            while(!server.isClosed()) {
                try {
                    final DatagramPacket packet = new DatagramPacket(new byte[1500], 1500);
                    server.receive(packet);
                    for(String msg : new String(packet.getData(), NonBlockingStatsDClient.MESSAGE_CHARSET).split("\n")) {
                        messagesReceived.add(msg.trim());
                    }
                } catch (IOException ignored) {}
            }
        });
        thread.setDaemon(true);
        thread.start();
    }

    public String nextMessage() throws InterruptedException{
        return messagesReceived.take();
    }

    public List<String> messagesReceived() {
        return new ArrayList<>(messagesReceived);
    }

    public void close() {
        server.close();
    }

    public void clear() {
        messagesReceived.clear();
    }
}


src/test/java/com/timgroup/statsd/DisruptorStatsDClientTest.java
package com.timgroup.statsd;

import org.junit.Before;

import java.net.ServerSocket;

import static org.junit.Assert.assertEquals;

public class DisruptorStatsDClientTest extends StringMessageStatsDClientTest {
    private String prefix;
    private int localPort;
    private DummyStatsDServer server;

    @Override
    @Before
    public void setUp() throws Exception {
        ServerSocket socket = new ServerSocket(0);
        int localPort = socket.getLocalPort();
        this.server = new DummyStatsDServer(localPort);
        this.localPort = localPort;
        this.prefix = "disruptor.prefix";
    }

    @Override
    protected StringMessageStatsDClient client() {
        return new DisruptorStatsDClient(prefix, "localhost", localPort);
    }

    @Override
    protected void assertRawMessageReceived(String expectedMessage) {
        try {
            assertEquals(expectedMessage, server.nextMessage());
        } catch (InterruptedException ignore) {}
    }

    @Override
    protected void assertMessageReceived(String expectedMessage) {
        assertRawMessageReceived(prefix + "." + expectedMessage);
    }

    @Override
    public void recordEvent() throws Exception {
        // we need to override this test from the parent test class due to the prefix
        Event event = Event.builder()
                .withTitle("title1")
                .withText("text1\nline2")
                .withDate(1234567000)
                .withHostname("host1")
                .withPriority(Event.Priority.LOW)
                .withAggregationKey("key1")
                .withAlertType(Event.AlertType.ERROR)
                .build();

        client().recordEvent(event);
        assertRawMessageReceived("_e{23,12}:disruptor.prefix.title1|text1\\nline2|d:1234567|h:host1|k:key1|p:low|t:error");

        client().recordEvent(event, "tag-1", "tag-2");
        assertRawMessageReceived("_e{23,12}:disruptor.prefix.title1|text1\\nline2|d:1234567|h:host1|k:key1|p:low|t:error|#tag-1,tag-2");
    }
}


src/main/java/com/timgroup/statsd/NoOpStatsDClient.java
package com.timgroup.statsd;

/**
 * A No-Op StatsDClient, which can be substituted in when metrics are not
 * required.
 *
 * @author Tom Denley
 *
 */
public class NoOpStatsDClient implements StatsDClient {
    @Override
    public void close() {}

    @Override
    public void count(String aspect, long delta, String... tags) {}

    @Override
    public void count(String aspect, long delta, double sampleRate, String... tags) {}

    @Override
    public void recordGaugeValue(String aspect, double value, String... tags) {}

    @Override
    public void recordGaugeValue(String aspect, double value, double sampleRate, String... tags) {}

    @Override
    public void recordGaugeValue(String aspect, long value, String... tags) {}

    @Override
    public void recordGaugeValue(String aspect, long value, double sampleRate, String... tags) {}

    @Override
    public void recordExecutionTime(String aspect, long timeInMs, String... tags) {}

    @Override
    public void recordExecutionTime(String aspect, long timeInMs, double sampleRate, String... tags) {}

    @Override
    public void recordHistogramValue(String aspect, double value, String... tags) {}

    @Override
    public void recordHistogramValue(String aspect, double value, double sampleRate, String... tags) {}

    @Override
    public void recordHistogramValue(String aspect, long value, String... tags) {}

    @Override
    public void recordHistogramValue(String aspect, long value, double sampleRate, String... tags) {}

    @Override
    public void recordEvent(Event event, String... tags) {}

    @Override
    public void recordServiceCheckRun(ServiceCheck sc) {}

    @Override
    public void recordSetValue(String aspect, String value, String... tags) {}
}


src/main/java/com/timgroup/statsd/DisruptorStatsDClient.java
package com.timgroup.statsd;

import com.lmax.disruptor.*;
import com.lmax.disruptor.dsl.Disruptor;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.DatagramChannel;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.TimeUnit;

/**
 * Client implementation that queues up metric messages in a LMAX disruptor.
 */
public class DisruptorStatsDClient extends StringMessageStatsDClient {

    public static final Charset MESSAGE_CHARSET = StandardCharsets.UTF_8;
    private static final int PACKET_SIZE_BYTES = 1400;
    private static final StatsDClientErrorHandler NO_OP_HANDLER = e -> {};
    private static final EventFactory<DisruptorEvent> FACTORY = DisruptorEvent::new;
    private static final EventTranslatorOneArg<DisruptorEvent, String> TRANSLATOR =
            (event, sequence, msg) -> event.setValue(msg);

    private final String prefix;
    private final DatagramChannel clientChannel;
    private final InetSocketAddress address;
    private final StatsDClientErrorHandler errorHandler;
    private final String constantTagsRendered;

    private final ThreadFactory threadFactory = r -> {
        Thread thread = new Thread(r);
        thread.setName("StatsD-disruptor-" + thread.getName());
        thread.setDaemon(true);
        return thread;
    };

    private final Disruptor<DisruptorEvent> disruptor = new Disruptor<>(FACTORY, 16384, threadFactory);

    public DisruptorStatsDClient(String prefix, String hostname, int port, String[] constantTags,
                                 StatsDClientErrorHandler errorHandler, DisruptorEventHandler handler)
            throws StatsDClientException {

        if (prefix != null && prefix.length() > 0) {
            this.prefix = String.format("%s.", prefix);
        } else {
            this.prefix = "";
        }

        this.errorHandler = errorHandler;

        if (constantTags != null && constantTags.length > 0) {
            this.constantTagsRendered = tagString(constantTags, null);
        } else {
            this.constantTagsRendered = null;
        }

        try {
            this.clientChannel = DatagramChannel.open();
            this.address = new InetSocketAddress(hostname, port);
        } catch (Exception e) {
            throw new StatsDClientException("Failed to start StatsD client", e);
        }

        disruptor.setDefaultExceptionHandler(new DisruptorExceptionHandler(this.errorHandler));

        if (handler != null) {
            disruptor.handleEventsWith(handler);
        } else {
            disruptor.handleEventsWith(new Handler());
        }

        disruptor.start();
    }

    public DisruptorStatsDClient(String prefix, String hostname, int port) throws StatsDClientException {
        this(prefix, hostname, port, null, NO_OP_HANDLER);
    }
    public DisruptorStatsDClient(String prefix, String hostname, int port, String[] constantTags)
            throws StatsDClientException {

        this(prefix, hostname, port, constantTags, NO_OP_HANDLER);
    }
    public DisruptorStatsDClient(String prefix, String hostname, int port, String[] constantTags,
                                 StatsDClientErrorHandler errorHandler) throws StatsDClientException {

        this(prefix, hostname, port, constantTags, errorHandler, null);
    }

    @Override
    public void close() {
        try {
            disruptor.shutdown(30, TimeUnit.SECONDS);
        } catch (Exception e) {
            errorHandler.handle(e);
        } finally {
            if (clientChannel != null) {
                try {
                    clientChannel.close();
                } catch (IOException e) {
                    errorHandler.handle(e);
                }
            }
        }
    }

    @Override
    public String getConstantTagsRendered() {
        return constantTagsRendered;
    }

    @Override
    public String getPrefix() {
        return prefix;
    }

    @Override
    protected void send(String message) {
        if (!disruptor.getRingBuffer().tryPublishEvent(TRANSLATOR, message)) {
            errorHandler.handle(InsufficientCapacityException.INSTANCE);
        }
    }

    public static class DisruptorEvent {
        private String value;

        public void setValue(String value) {
            this.value = value;
        }

        @Override
        public String toString() {
            return "Event: " + value;
        }
    }

    /**
     * Avoids "unchecked generic array creation for varargs parameter" errors when using the disruptor library
     */
    public interface DisruptorEventHandler extends EventHandler<DisruptorEvent> {}

    protected class Handler implements DisruptorEventHandler {
        private final ByteBuffer sendBuffer = ByteBuffer.allocate(PACKET_SIZE_BYTES);

        @Override
        public void onEvent(DisruptorEvent event, long sequence, boolean batchEnd) throws Exception {
            String message = event.value;
            byte[] data = message.getBytes(MESSAGE_CHARSET);
            if (sendBuffer.remaining() < (data.length + 1)) {
                flush();
            }
            if (sendBuffer.position() > 0) {
                sendBuffer.put((byte) '\n');
            }
            sendBuffer.put(
                    data.length > sendBuffer.remaining() ? Arrays.copyOfRange(data, 0, sendBuffer.remaining()) : data);

            if (batchEnd || 0 == sendBuffer.remaining()) {
                flush();
            }
        }

        private void flush() throws IOException {
            int sizeOfBuffer = sendBuffer.position();
            sendBuffer.flip();
            int sentBytes = clientChannel.send(sendBuffer, address);
            sendBuffer.clear();

            if (sizeOfBuffer != sentBytes) {
                errorHandler.handle(new IOException(
                        String.format(
                                "Could not send entirely stat %s to host %s:%d. Only sent %d bytes out of %d bytes",
                                sendBuffer.toString(),
                                address.getHostName(),
                                address.getPort(),
                                sentBytes,
                                sizeOfBuffer)));
            }
        }
    }

    protected static class DisruptorExceptionHandler implements ExceptionHandler<Object> {
        private final FatalExceptionHandler throwableHandler = new FatalExceptionHandler();
        private final StatsDClientErrorHandler exceptionHandler;

        public DisruptorExceptionHandler(StatsDClientErrorHandler handler) {
            this.exceptionHandler = handler;
        }

        @Override
        public void handleEventException(Throwable ex, long sequence, Object event) {
            if (ex instanceof Exception) {
                exceptionHandler.handle((Exception) ex);
            } else {
                throwableHandler.handleEventException(ex, sequence, event);
            }
        }

        @Override
        public void handleOnStartException(Throwable ex) {
            if (ex instanceof Exception) {
                exceptionHandler.handle((Exception) ex);
            } else {
                throwableHandler.handleOnStartException(ex);
            }
        }

        @Override
        public void handleOnShutdownException(Throwable ex) {
            if (ex instanceof Exception) {
                exceptionHandler.handle((Exception) ex);
            } else {
                throwableHandler.handleOnShutdownException(ex);
            }
        }
    }
}


src/main/java/com/timgroup/statsd/NonBlockingStatsDClient.java
package com.timgroup.statsd;

import java.io.IOException;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.UnknownHostException;
import java.nio.ByteBuffer;
import java.nio.channels.DatagramChannel;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.*;



/**
 * A simple StatsD client implementation facilitating metrics recording.
 *
 * <p>Upon instantiation, this client will establish a socket connection to a StatsD instance
 * running on the specified host and port. Metrics are then sent over this connection as they are
 * received by the client.
 * </p>
 *
 * <p>Three key methods are provided for the submission of data-points for the application under
 * scrutiny:
 * <ul>
 *   <li>{@link #incrementCounter} - adds one to the value of the specified named counter</li>
 *   <li>{@link #recordGaugeValue} - records the latest fixed value for the specified named gauge</li>
 *   <li>{@link #recordExecutionTime} - records an execution time in milliseconds for the specified named operation</li>
 *   <li>{@link #recordHistogramValue} - records a value, to be tracked with average, maximum, and percentiles</li>
 *   <li>{@link #recordEvent} - records an event</li>
 *   <li>{@link #recordSetValue} - records a value in a set</li>
 * </ul>
 * From the perspective of the application, these methods are non-blocking, with the resulting
 * IO operations being carried out in a separate thread. Furthermore, these methods are guaranteed
 * not to throw an exception which may disrupt application execution.
 *
 * <p>As part of a clean system shutdown, the {@link #close()} method should be invoked
 * on any StatsD clients.</p>
 *
 * @author Tom Denley
 *
 */
public class NonBlockingStatsDClient extends StringMessageStatsDClient {

    public static final Charset MESSAGE_CHARSET = StandardCharsets.UTF_8;
    private static final int PACKET_SIZE_BYTES = 1400;
    private static final StatsDClientErrorHandler NO_OP_HANDLER = e -> { /* No-op */ };

    private final String prefix;
    private final DatagramChannel clientChannel;
    private final StatsDClientErrorHandler handler;
    private final String constantTagsRendered;

    private final ExecutorService executor = Executors.newSingleThreadExecutor(new ThreadFactory() {
        final ThreadFactory delegate = Executors.defaultThreadFactory();
        @Override public Thread newThread(final Runnable r) {
            final Thread result = delegate.newThread(r);
            result.setName("StatsD-" + result.getName());
            result.setDaemon(true);
            return result;
        }
    });

    private final BlockingQueue<String> queue;

    /**
     * Create a new StatsD client communicating with a StatsD instance on the
     * specified host and port. All messages send via this client will have
     * their keys prefixed with the specified string. The new client will
     * attempt to open a connection to the StatsD server immediately upon
     * instantiation, and may throw an exception if that a connection cannot
     * be established. Once a client has been instantiated in this way, all
     * exceptions thrown during subsequent usage are consumed, guaranteeing
     * that failures in metrics will not affect normal code execution.
     *
     * @param prefix
     *     the prefix to apply to keys sent via this client
     * @param hostname
     *     the host name of the targeted StatsD server
     * @param port
     *     the port of the targeted StatsD server
     * @throws StatsDClientException
     *     if the client could not be started
     */
    public NonBlockingStatsDClient(final String prefix, final String hostname, final int port) throws StatsDClientException {
        this(prefix, hostname, port, Integer.MAX_VALUE);
    }

    /**
     * Create a new StatsD client communicating with a StatsD instance on the
     * specified host and port. All messages send via this client will have
     * their keys prefixed with the specified string. The new client will
     * attempt to open a connection to the StatsD server immediately upon
     * instantiation, and may throw an exception if that a connection cannot
     * be established. Once a client has been instantiated in this way, all
     * exceptions thrown during subsequent usage are consumed, guaranteeing
     * that failures in metrics will not affect normal code execution.
     *
     * @param prefix
     *     the prefix to apply to keys sent via this client
     * @param hostname
     *     the host name of the targeted StatsD server
     * @param port
     *     the port of the targeted StatsD server
     * @param queueSize
     *     the maximum amount of unprocessed messages in the BlockingQueue.
     * @throws StatsDClientException
     *     if the client could not be started
     */
    public NonBlockingStatsDClient(final String prefix, final String hostname, final int port, final int queueSize) throws StatsDClientException {
        this(prefix, hostname, port, queueSize, null, null);
    }

    /**
     * Create a new StatsD client communicating with a StatsD instance on the
     * specified host and port. All messages send via this client will have
     * their keys prefixed with the specified string. The new client will
     * attempt to open a connection to the StatsD server immediately upon
     * instantiation, and may throw an exception if that a connection cannot
     * be established. Once a client has been instantiated in this way, all
     * exceptions thrown during subsequent usage are consumed, guaranteeing
     * that failures in metrics will not affect normal code execution.
     *
     * @param prefix
     *     the prefix to apply to keys sent via this client
     * @param hostname
     *     the host name of the targeted StatsD server
     * @param port
     *     the port of the targeted StatsD server
     * @param constantTags
     *     tags to be added to all content sent
     * @throws StatsDClientException
     *     if the client could not be started
     */
    public NonBlockingStatsDClient(final String prefix, final String hostname, final int port, final String... constantTags) throws StatsDClientException {
        this(prefix, hostname, port, Integer.MAX_VALUE, constantTags, null);
    }

    /**
     * Create a new StatsD client communicating with a StatsD instance on the
     * specified host and port. All messages send via this client will have
     * their keys prefixed with the specified string. The new client will
     * attempt to open a connection to the StatsD server immediately upon
     * instantiation, and may throw an exception if that a connection cannot
     * be established. Once a client has been instantiated in this way, all
     * exceptions thrown during subsequent usage are consumed, guaranteeing
     * that failures in metrics will not affect normal code execution.
     *
     * @param prefix
     *     the prefix to apply to keys sent via this client
     * @param hostname
     *     the host name of the targeted StatsD server
     * @param port
     *     the port of the targeted StatsD server
     * @param constantTags
     *     tags to be added to all content sent
     * @param queueSize
     *     the maximum amount of unprocessed messages in the BlockingQueue.
     * @throws StatsDClientException
     *     if the client could not be started
     */
    public NonBlockingStatsDClient(final String prefix, final String hostname, final int port, final int queueSize, final String... constantTags) throws StatsDClientException {
        this(prefix, hostname, port, queueSize, constantTags, null);
    }

    /**
     * Create a new StatsD client communicating with a StatsD instance on the
     * specified host and port. All messages send via this client will have
     * their keys prefixed with the specified string. The new client will
     * attempt to open a connection to the StatsD server immediately upon
     * instantiation, and may throw an exception if that a connection cannot
     * be established. Once a client has been instantiated in this way, all
     * exceptions thrown during subsequent usage are passed to the specified
     * handler and then consumed, guaranteeing that failures in metrics will
     * not affect normal code execution.
     *
     * @param prefix
     *     the prefix to apply to keys sent via this client
     * @param hostname
     *     the host name of the targeted StatsD server
     * @param port
     *     the port of the targeted StatsD server
     * @param constantTags
     *     tags to be added to all content sent
     * @param errorHandler
     *     handler to use when an exception occurs during usage, may be null to indicate noop
     * @throws StatsDClientException
     *     if the client could not be started
     */
    public NonBlockingStatsDClient(final String prefix, final String hostname, final int port,
                                   final String[] constantTags, final StatsDClientErrorHandler errorHandler) throws StatsDClientException {
        this(prefix, Integer.MAX_VALUE, constantTags, errorHandler, staticStatsDAddressResolution(hostname, port));
    }

    /**
     * Create a new StatsD client communicating with a StatsD instance on the
     * specified host and port. All messages send via this client will have
     * their keys prefixed with the specified string. The new client will
     * attempt to open a connection to the StatsD server immediately upon
     * instantiation, and may throw an exception if that a connection cannot
     * be established. Once a client has been instantiated in this way, all
     * exceptions thrown during subsequent usage are passed to the specified
     * handler and then consumed, guaranteeing that failures in metrics will
     * not affect normal code execution.
     *
     * @param prefix
     *     the prefix to apply to keys sent via this client
     * @param hostname
     *     the host name of the targeted StatsD server
     * @param port
     *     the port of the targeted StatsD server
     * @param constantTags
     *     tags to be added to all content sent
     * @param errorHandler
     *     handler to use when an exception occurs during usage, may be null to indicate noop
     * @param queueSize
     *     the maximum amount of unprocessed messages in the BlockingQueue.
     * @throws StatsDClientException
     *     if the client could not be started
     */
    public NonBlockingStatsDClient(final String prefix, final String hostname, final int port, final int queueSize,
                                   final String[] constantTags, final StatsDClientErrorHandler errorHandler) throws StatsDClientException {
        this(prefix, queueSize, constantTags, errorHandler, staticStatsDAddressResolution(hostname, port));
    }

    /**
     * Create a new StatsD client communicating with a StatsD instance on the
     * specified host and port. All messages send via this client will have
     * their keys prefixed with the specified string. The new client will
     * attempt to open a connection to the StatsD server immediately upon
     * instantiation, and may throw an exception if that a connection cannot
     * be established. Once a client has been instantiated in this way, all
     * exceptions thrown during subsequent usage are passed to the specified
     * handler and then consumed, guaranteeing that failures in metrics will
     * not affect normal code execution.
     *
     * @param prefix
     *     the prefix to apply to keys sent via this client
     * @param constantTags
     *     tags to be added to all content sent
     * @param errorHandler
     *     handler to use when an exception occurs during usage, may be null to indicate noop
     * @param addressLookup
     *     yields the IP address and socket of the StatsD server
     * @param queueSize
     *     the maximum amount of unprocessed messages in the BlockingQueue.
     * @throws StatsDClientException
     *     if the client could not be started
     */
    public NonBlockingStatsDClient(final String prefix, final int queueSize, String[] constantTags, final StatsDClientErrorHandler errorHandler,
                                   final Callable<InetSocketAddress> addressLookup) throws StatsDClientException {
        if((prefix != null) && (!prefix.isEmpty())) {
            this.prefix = String.format("%s.", prefix);
        } else {
            this.prefix = "";
        }
        if(errorHandler == null) {
            handler = NO_OP_HANDLER;
        }
        else {
            handler = errorHandler;
        }

        /* Empty list should be null for faster comparison */
        if((constantTags != null) && (constantTags.length == 0)) {
            constantTags = null;
        }

        if(constantTags != null) {
            constantTagsRendered = tagString(constantTags, null);
        } else {
            constantTagsRendered = null;
        }

        try {
            clientChannel = DatagramChannel.open();
        } catch (final Exception e) {
            throw new StatsDClientException("Failed to start StatsD client", e);
        }
        queue = new LinkedBlockingQueue<>(queueSize);

        executor.execute(new QueueConsumer(addressLookup));
    }

    /**
     * Cleanly shut down this StatsD client. This method may throw an exception if
     * the socket cannot be closed.
     */
    @Override
    public void close() {
        try {
            executor.shutdown();
            executor.awaitTermination(30, TimeUnit.SECONDS);
        }
        catch (final Exception e) {
            handler.handle(e);
        }
        finally {
            if (clientChannel != null) {
                try {
                    clientChannel.close();
                }
                catch (final IOException e) {
                    handler.handle(e);
                }
            }
        }
    }

    @Override
    protected void send(final String message) {
        queue.offer(message);
    }

    @Override
    public String getConstantTagsRendered() {
        return constantTagsRendered;
    }

    @Override
    public String getPrefix() {
        return prefix;
    }

    private class QueueConsumer implements Runnable {
        private final ByteBuffer sendBuffer = ByteBuffer.allocate(PACKET_SIZE_BYTES);

        private final Callable<InetSocketAddress> addressLookup;

        QueueConsumer(final Callable<InetSocketAddress> addressLookup) {
            this.addressLookup = addressLookup;
        }

        @Override public void run() {
            while(!executor.isShutdown()) {
                try {
                    final String message = queue.poll(1, TimeUnit.SECONDS);
                    if(null != message) {
                        final InetSocketAddress address = addressLookup.call();
                        final byte[] data = message.getBytes(MESSAGE_CHARSET);
                        if(sendBuffer.remaining() < (data.length + 1)) {
                            blockingSend(address);
                        }
                        if(sendBuffer.position() > 0) {
                            sendBuffer.put( (byte) '\n');
                        }
                        sendBuffer.put(data);
                        if(null == queue.peek()) {
                            blockingSend(address);
                        }
                    }
                } catch (final Exception e) {
                    handler.handle(e);
                }
            }
        }

        private void blockingSend(final InetSocketAddress address) throws IOException {
            final int sizeOfBuffer = sendBuffer.position();
            sendBuffer.flip();

            final int sentBytes = clientChannel.send(sendBuffer, address);
            sendBuffer.limit(sendBuffer.capacity());
            sendBuffer.rewind();

            if (sizeOfBuffer != sentBytes) {
                handler.handle(
                        new IOException(
                            String.format(
                                "Could not send entirely stat %s to host %s:%d. Only sent %d bytes out of %d bytes",
                                sendBuffer.toString(),
                                address.getHostName(),
                                address.getPort(),
                                sentBytes,
                                sizeOfBuffer)));
            }
        }
    }

    /**
     * Create dynamic lookup for the given host name and port.
     *
     * @param hostname the host name of the targeted StatsD server
     * @param port     the port of the targeted StatsD server
     * @return a function to perform the lookup
     */
    public static Callable<InetSocketAddress> volatileAddressResolution(final String hostname, final int port) {
        return () -> new InetSocketAddress(InetAddress.getByName(hostname), port);
    }

  /**
   * Lookup the address for the given host name and cache the result.
   *
   * @param hostname the host name of the targeted StatsD server
   * @param port     the port of the targeted StatsD server
   * @return a function that cached the result of the lookup
   * @throws Exception if the lookup fails, i.e. {@link UnknownHostException}
   */
    public static Callable<InetSocketAddress> staticAddressResolution(final String hostname, final int port) throws Exception {
        final InetSocketAddress address = volatileAddressResolution(hostname, port).call();
        return () -> address;
    }

    private static Callable<InetSocketAddress> staticStatsDAddressResolution(final String hostname, final int port) throws StatsDClientException {
        try {
            return staticAddressResolution(hostname, port);
        } catch (final Exception e) {
            throw new StatsDClientException("Failed to lookup StatsD host", e);
        }
    }
}
